### PR TITLE
Use index scale instead of rank scale + color scale and custom transf…

### PR DESCRIPTION
**Pull request by domoritz**
```json
{
  "data": {
    "values": [
      {"a": "A","b": 28, "c": "x-small"},
      {"a": "B","b": 55, "c": "small"},
      {"a": "C","b": 43, "c": "x-large"},
      {"a": "D","b": 13, "c": "large"}
    ]
  },
  "mark": "bar",
  "encoding": {
    "x": {"field": "a", "type": "ordinal"},
    "y": {"field": "b", "type": "quantitative"},
    "color": {"field": "c", "type": "ordinal"}
  }
}
```

screen shot 2016-12-01 at 22 31 24

### CODE REVIEW

**Comment by domoritz** [comment]
I also changed the default color scale to viridis. If you want to have some more discussion before we do that, I am happy to delete the commit.

**Comment by domoritz** [comment]
Fixes https://github.com/vega/vega-lite/issues/1732

File: src/compile/scale.ts [hunk 1]
```diff
@@ -348,6 +315,9 @@ export namespace defaultProperty {
         return ScaleType.TIME;
 
       case QUANTITATIVE:
+      if (fieldDef.bin) {
```
Comment (kanitw) [inline src/compile/scale.ts]: wrong tab :)

File: src/scale.ts [hunk 2]
```diff
@@ -22,14 +22,17 @@ export namespace ScaleType {
   export const ORDINAL_LOOKUP: 'ordinal' = 'ordinal';
   export const POINT: 'point' = 'point';
   export const BAND: 'band' = 'band';
+
+  // Ordinal color scale
+  export const INDEX: 'index' = 'index';
```
Comment (kanitw) [inline src/scale.ts]: Need to update `scaleTypeSupportProperty` at the end of this file and write test.

File: src/compile/legend.ts [hunk 3]
```diff
@@ -228,24 +209,17 @@ export namespace encode {
...
-            field: 'data'
-          }
-        }, labelsSpec || {});
-      } else if (fieldDef.bin) {
+      if (fieldDef.bin) {
         labelsSpec = extend({
           text: {
             scale: model.scaleName(COLOR_LEGEND_LABEL),
-            field: 'data'
+            field: 'value'
```
Comment (kanitw) [inline src/compile/legend.ts]: i guess we need **another** follow up PR to change data=>value for all axis/legend stuff?  (File an issue?)

File: src/mark.ts [hunk 4]
```diff
@@ -190,7 +190,7 @@ export interface MarkConfig {
 export const defaultMarkConfig: MarkConfig = {
   color: '#4682b4',
   nominalColorScheme: 'category10',
-  sequentialColorScheme: 'Greens',
+  sequentialColorScheme: 'viridis',
```
Comment (kanitw) [inline src/mark.ts]: Maybe this is better, but let's not worry about this for now.

File: src/compile/scale.ts [hunk 5]
```diff
@@ -182,11 +174,8 @@ export function parseScaleComponent(model: Model): Dict<ScaleComponents> {
 
         // Add additional scales needed to support ordinal legends (list of values)
         // for color ramp.
-        if (channel === COLOR && model.legend(COLOR) && (fieldDef.type === ORDINAL || fieldDef.bin || fieldDef.timeUnit)) {
-          scales.colorLegend = parseColorLegendScale(model, fieldDef);
-          if (fieldDef.bin) {
-            scales.binColorLegend = parseBinColorLegendLabel(model, fieldDef);
-          }
+        if (channel === COLOR && model.legend(COLOR) && fieldDef.bin) {
```
Comment (kanitw) [inline src/compile/scale.ts]: The comment above is outdated.  Now you only do it for bin. 

(why? -- add a comment why you need it for bin too, right now I don't understand why you still need it for bin.)

File: src/compile/scale.ts (hunk 1 above)
Comment (domoritz) [inline src/compile/scale.ts]: Done

File: src/compile/legend.ts (hunk 3 above)
Comment (domoritz) [inline src/compile/legend.ts]: Will come across these later.

File: src/mark.ts (hunk 4 above)
Comment (domoritz) [inline src/mark.ts]: Okay. Removed from this pr.

File: src/compile/legend.ts (hunk 3 above)
Comment (kanitw) [inline src/compile/legend.ts]: Ok, I added a TODO [here](https://github.com/vega/vega-lite/issues/1597#issuecomment-255143163).

File: src/compile/scale.ts (hunk 5 above)
Comment (kanitw) [inline src/compile/scale.ts]: The comment is still outdated

Comment (domoritz) [inline src/compile/scale.ts]: Not pushed yet.

File: src/compile/scale.ts [hunk 6]
```diff
@@ -327,14 +296,12 @@ export namespace defaultProperty {
...
-          // TODO: check if this is still true
-          return ScaleType.LINEAR; // ordinal has order, so use interpolated ordinal color scale.
+          return ScaleType.INDEX;
         }
         return discreteToContinuousType(channel, mark, canHaveRangeStep);
       case TEMPORAL:
         if (channel === COLOR) {
-          // FIXME: or sequential?
-          return ScaleType.TIME; // time has order, so use interpolated color scale.
+          return ScaleType.INDEX;
```
Comment (kanitw) [inline src/compile/scale.ts]: and for both timeUnit that is time scale-based or ordinal scale-based

File: examples/specs/scatter_binned_opacity.vl.json [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{
+  "description": "A scatterplot showing horsepower and miles per gallons with binned acceleration on size.",
```
Comment (kanitw) [inline examples/specs/scatter_binned_opacity.vl.json]: on "opacity"

File: src/compile/legend.ts [hunk 8]
```diff
@@ -29,25 +30,18 @@ export function parseLegendComponent(model: UnitModel): Dict<VgLegend> {
 }
 
 function getLegendDefWithScale(model: UnitModel, channel: Channel): VgLegend {
+  // use a special scale so we can overrride the mark props and labels
```
Comment (kanitw) [inline src/compile/legend.ts]: "For binned field with continuous scale, use a ... "

(The current comment misleads that we always use...)

File: src/compile/legend.ts [hunk 9]
```diff
@@ -212,6 +196,24 @@ export namespace encode {
...
+        case OPACITY:
+          symbols.opacity = def;
+          break;
+        case SIZE:
+          symbols.size = def;
+          break;
+        case COLOR:
+          symbols[filled ? 'fill' : 'stroke'] = def;
+          break;
+      }
```
Comment (kanitw) [inline src/compile/legend.ts]: Let's add:

```
default: 
  throw Error(`Legend for channel ${channel} not implemented`);  // put this message in log.message
```

so when someone try to add new channels, they will know suddenly that something is wrong.

File: src/compile/legend.ts [hunk 10]
```diff
@@ -227,28 +229,19 @@ export namespace encode {
...
-        }, labelsSpec || {});
-      } else if (fieldDef.type === TEMPORAL) {
-        labelsSpec = extend({
-          text: {
-            signal: timeFormatExpression('datum["data"]', fieldDef.timeUnit, legend.format, legend.shortTimeLabels, config)
-          }
-        }, labelsSpec || {});
-      }
+    if (fieldDef.bin && isContinuousScale(model.scale(channel).type)) {
+      labelsSpec = extend({
```
Comment (kanitw) [inline src/compile/legend.ts]: // Override label's text to map bin's quantitative value to range

File: src/compile/model.ts [hunk 11]
```diff
@@ -341,7 +341,7 @@ export abstract class Model {
    * (DO NOT USE THIS METHOD DURING SCALE PARSING, use model.name() instead)
    */
   public scaleName(originalScaleName: Channel|string, parse?: boolean): string {
-    const channel = contains([COLOR_LEGEND, COLOR_LEGEND_LABEL], originalScaleName) ? 'color' : originalScaleName;
+    const channel = originalScaleName.replace(BIN_LEGEND_SUFFIX, '');
```
Comment (kanitw) [inline src/compile/model.ts]: How about BIN_LEGEND_LABEL_SUFFIX?

File: src/compile/scale.ts [hunk 12]
```diff
@@ -19,15 +19,10 @@ import {Model} from './model';
...
-/**
- * Color Ramp's scale for legends.  This scale has to be ordinal so that its
- * legends show a list of numbers.
- */
-export const COLOR_LEGEND = 'color_legend';
-
-// scale used to get labels for binned color scales
-export const COLOR_LEGEND_LABEL = 'color_legend_label';
-
+// scale used to get drive binned legends
```
Comment (kanitw) [inline src/compile/scale.ts]: /** Scale suffix for driving binned field legends. */

File: src/compile/scale.ts [hunk 13]
```diff
@@ -19,15 +19,10 @@ import {Model} from './model';
...
+export const BIN_LEGEND_SUFFIX = '_bin_legend';
+// scale used to get labels for legend for binned scales
```
Comment (kanitw) [inline src/compile/scale.ts]: /** Scale suffix for binned field's legend labels, which maps a binned field's quantitative values to range strings. */

File: src/compile/scale.ts [hunk 14]
```diff
@@ -507,34 +497,36 @@ export function domain(scale: Scale, model: Model, channel:Channel): any {
...
+        // X/Y position have to include start and end for non-ordinal scale
+        return {
+          data: model.dataTable(),
+          fields: [
+            model.field(channel, { binSuffix: 'start' }),
+            model.field(channel, { binSuffix: 'end' })
+          ]
+        };
+      } else {
+        // TODO: use bin_mid
```
Comment (kanitw) [inline src/compile/scale.ts]: File an issue to follow up?

File: src/scale.ts [hunk 15]
```diff
@@ -42,18 +45,18 @@ export const SCALE_TYPES: ScaleType[] = [
...
 const DISCRETE_SCALE_TYPE_INDEX = toSet(DISCRETE_SCALE_TYPES);
 
 export const TIME_SCALE_TYPES: ScaleType[] = ['time', 'utc'];
 
-export function isDiscreteScale(type: ScaleType): type is 'ordinal' | 'point' | 'band' {
+export function isDiscreteScale(type: ScaleType): type is 'ordinal' | 'point' | 'band' | 'index' {
   return type in DISCRETE_SCALE_TYPE_INDEX;
 }
-export function isContinuousScale(type: ScaleType): type is 'linear' | 'log' | 'pow' | 'sqrt' |  'time' | 'utc' {
+export function isContinuousScale(type: ScaleType): type is 'linear' | 'log' | 'pow' | 'sequential' | 'sqrt' |  'time' | 'utc' {
```
Comment (kanitw) [inline src/scale.ts]: This modification is wrong. Please merge #1755 into vg3 and use `hasContinuousDomain` in the new predicates we add in this PR.

**Comment by domoritz** [comment]
Rebased

File: src/compile/data/bin.ts [hunk 16]
```diff
@@ -46,11 +47,9 @@ export namespace bin {
 
         transform.push(binTrans);
 
-        // If color ramp has type linear or time, we have to create new bin_range field
-        // with correct number format
-        const isOrdinalColor = model.hasDiscreteScale(channel) || channel === COLOR;
-        if (isOrdinalColor) {
-          // read format from axis or legend, if there is not format there then use config.numberFormat
+        const isDiscreteOrHasLegend = hasDiscreteDomain(model.scale(channel).type) || model.legend(channel);
```
Comment (kanitw) [inline src/compile/data/bin.ts]: hasDiscreteDomainOrHasLegend?

File: src/compile/scale.ts [hunk 17]
```diff
@@ -19,15 +19,10 @@ import {Model} from './model';
...
-/**
- * Color Ramp's scale for legends.  This scale has to be ordinal so that its
- * legends show a list of numbers.
- */
-export const COLOR_LEGEND = 'color_legend';
-
-// scale used to get labels for binned color scales
-export const COLOR_LEGEND_LABEL = 'color_legend_label';
-
+// Scale suffix for scale used to get drive binned legends.
```
Comment (kanitw) [inline src/compile/scale.ts]: Why not /** */ ? 

If you use /** */, you can mouseover and get tooltip in VSCode too.  :)

**Comment by kanitw** [comment]
Besides /** */, this LGTM -- feel free to merge after fixing. 

I assume that you have thoroughly revised `scaleTypeSupportProperty` -- I read through it and can't think of anything incorrect. The only weird thing is that `index` support `range` since it's equivalent to `point` scale in that case.  However, this matches our experiment yesterday and Vega's JSON schema.

**Comment by kanitw** [comment]
🎉

**Pull request merged**